dawn/prototype/AtlasFromNetcdf.cpp

```diff
@@ -0,0 +1,273 @@
+                                 cellToVertex[2 * ncells + cellIdx] - 1};
+    node_connectivity.set(cellIdx, tri_nodes);
+    glb_idx_cell[cellIdx] = cellIdx;
+    cells_part(cellIdx) = defaultPartition;
+  }
+
+  return true;
+}
+
+template <typename ConnectivityT>
+bool AddNeighborList(const netCDF::NcFile& dataFile, atlas::Mesh& mesh,
+                     const std::string nbhListName, size_t yPerXExpected,
+                     ConnectivityT& connectivity) {
+  auto [xToY, yPerX, numY] = Load2DField<int>(dataFile, nbhListName);
+  if(yPerX != yPerXExpected) {
+    std::cout << "number of neighbors per element not as expected!\n";
+    return false;
+  }
+  for(size_t elemIdx = 0; elemIdx < numY; elemIdx++) {
+    atlas::idx_t yOfX[yPerX];
+    for(size_t innerIdx = 0; innerIdx < yPerX; innerIdx++) {
+      // indices in netcdf are 1 based, data is column major
+      yOfX[innerIdx] = xToY[innerIdx * numY + elemIdx] - 1;
+    }
+    connectivity.set(elemIdx, yOfX);
+  }
+  return true;
+}
+
+// Pre-allocate neighborhood table, snippet taken from
+// https://github.com/ecmwf/atlas/blob/98ff1f20dc883ba2dfd7658196622b9bc6d6fceb/src/atlas/mesh/actions/BuildEdges.cc#L73
+template <typename ConnectivityT>
+void AllocNbhTable(ConnectivityT& connectivity, int numElements, int nbhPerElem) {
+  std::vector<int> init(numElements * nbhPerElem, connectivity.missing_value());
+  connectivity.add(numElements, nbhPerElem, init.data());
+}
+} // namespace
+
+std::optional<atlas::Mesh> AtlasMeshFromNetCDFMinimal(const std::string& filename) {
+  try {
+    atlas::Mesh mesh;
+    netCDF::NcFile dataFile(filename.c_str(), netCDF::NcFile::read);
+
+    if(!NodesFromNetCDF(dataFile, mesh)) {
+      return {};
+    }
+
+    if(!CellsFromNetCDF(dataFile, mesh)) {
+      return {};
+    }
+
+    return mesh;
+  } catch(netCDF::exceptions::NcException& e) {
+    std::cout << e.what() << "\n";
+    return std::nullopt;
+  }
+}
+
+std::optional<atlas::Mesh> AtlasMeshFromNetCDFComplete(const std::string& filename) {
+  auto maybeMesh = AtlasMeshFromNetCDFMinimal(filename);
+  if(!maybeMesh.has_value()) {
+    return {};
+  }
+
+  try {
+    atlas::Mesh mesh = maybeMesh.value();
+    netCDF::NcFile dataFile(filename.c_str(), netCDF::NcFile::read);
+
+    int numEdges = LoadField<int>(dataFile, "edge_index").size();
+    if(numEdges == 0) {
+      std::cout << "no edges found in netcdf file!\n";
+      return {};
+    }
+
+    // had no edges so far, add them
+    mesh.edges().add(new atlas::mesh::temporary::Line(), numEdges);
+
+    const int verticesPerEdge = 2;
+    const int cellsPerEdge = 2;
+    const int cellsPerNode = 6; // maximum is 6, some with 5 exist
+    const int edgesPerNode = 6; // maximum is 6, some with 5 exist
+    const int edgesPerCell = 3;
+
+    // Allocate & fill neighbor tables from file
+    // ------------------------------
+
+    // Edges
+    AllocNbhTable<atlas::mesh::HybridElements::Connectivity>(mesh.edges().cell_connectivity(),
```

Comment:
Why do you need to explicitly specify the template argument? What type is `mesh.edges().cell_connectivity()`?

Reply:
I realize I don't, I was simply not used to doing it this way, thanks.